src/extensions/agent_modules/cloud_communication/adps2_mqtt_client/src/adps2_mqtt_client_module.c

```diff
@@ -0,0 +1,626 @@
+    - The device must poll the service periodically to receive the result of the device registration operation.
+    - Subscribe to the $dps/registrations/res/# topic
+    - Publish a get operation status message to $dps/registrations/GET/iotdps-get-operationstatus/?$rid={request_id}&operationId={operationId}
+        - The operation ID in this message should be the value received in the RegistrationOperationStatus
+          response message in the previous step.
+    - In the successful case, the service responds on the $dps/registrations/res/200/?$rid={request_id} topic.
+    - The payload of the response contains the RegistrationOperationStatus object.
+    - The device should keep polling the service if the response code is 202 after a delay equal to the retry-after period.
+    - The device registration operation is successful if the service returns a 200 status code.
+*/
+
+bool DeviceRegistration_DoWork(/*const char* request_id, const char* connect_json*/)
+{
+    if (s_moduleState.registerState == ADPS_REGISTER_STATE_REGISTERED)
+    {
+        return true;
+    }
+
+    time_t nowTime = ADUC_GetTimeSinceEpochInSeconds();
+
+    if (s_moduleState.registerState == ADPS_REGISTER_STATE_REGISTERING)
+    {
+        if (nowTime > s_moduleState.nextRegisterAttemptTime)
+        {
+            SetRegisterState(ADPS_REGISTER_STATE_UNKNOWN, "time out");
+        }
+    }
+
+    if (s_moduleState.registerState == ADPS_REGISTER_STATE_WAIT_TO_POLL)
+    {
+        if (nowTime > s_moduleState.nextRegisterAttemptTime)
+        {
+            char* pollTopic = ADUC_StringFormat(
+                "$dps/registrations/GET/iotdps-get-operationstatus/?$rid=%s&operationId=%s",
+                s_moduleState.settings.registrationId,
+                s_moduleState.operationId);
+            if (pollTopic == NULL)
+            {
+                Log_Error("Cannot allocate memory for registration status polling topic");
+            }
+            else
+            {
+                int mid = 0;
+                int mqtt_ret = ADUC_CommunicationChannel_MQTT_Publish(
+                    pollTopic,
+                    &mid,
+                    0,
+                    NULL,
+                    s_moduleState.settings.mqttSettings.qos,
+                    false /* retain */,
+                    NULL /* props */);
+
+                if (mqtt_ret != MOSQ_ERR_SUCCESS)
+                {
+                    Log_Error("Failed to publish registration polling request (%d)", mqtt_ret);
+                    // TODO (nox-msft) : determine whether to retry or bail
+                }
+                else
+                {
+                    SetRegisterState(ADPS_REGISTER_STATE_POLLING, "publishing message");
+                    s_moduleState.nextRegisterAttemptTime = nowTime + ADPS_DEFAULT_REGISTER_REQUEST_DELAY_SECONDS;
+                    s_moduleState.registerRetries++;
+                }
+            }
+            free(pollTopic);
```

Comment:
not doing free at done, indicates that the body of the `if (s_moduleState.registerState == ADPS_REGISTER_STATE_WAIT_TO_POLL)` condition should be factored out into its own function, e.g. `handle_wait_to_poll_register_state()`. once implementing retry/bail logic, it's prone to error and forgetting leak memory by skipping the free.